craft_store/publishergateway/_request.py

```diff
@@ -0,0 +1,41 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Request models for the publisher gateway."""
+
+import re
+from typing import Annotated, TypedDict
+
+import annotated_types
+from typing_extensions import NotRequired
+
+TRACK_NAME_REGEX = re.compile(r"^[a-zA-Z0-9](?:[_.-]?[a-zA-Z0-9])*$")
+"""A regular expression guarding track names.
+
+Retrieved from https://api.staging.charmhub.io/docs/default.html#create_tracks
+"""
+
+CreateTrackRequest = TypedDict(
```

Comment:
We're trying to move away from generating and transforming the models all the time. This allows us to return the JSON dictionary returned by the store but still provide users with helpful details and linting during development.

At runtime, these are a no-op, but during development they offer both this:

![Screenshot of kate with the pyright LSP completing possible package metadata keys starting with a t](https://github.com/user-attachments/assets/a47f187c-36ac-417f-a1a5-3ecea34b17b9)

and this:

![image](https://github.com/user-attachments/assets/0bb6070c-d56c-4224-8e3b-497d13c218d0)

Reply:
I won't block you from delivering that, but for me it looks terrible and is really bug-prone - no validation at runtime unless you explicitly do it by yourself. No control over values returned from the store and 

At least we should use a `dataclass`-like regular syntax to make it somehow readable. 
https://docs.python.org/3.12/library/typing.html#typing.TypedDict

We also should at some point enforce using `TypeAdapter`s for validating all the data that gets in and out:
https://docs.pydantic.dev/latest/concepts/type_adapter/#__tabbed_1_2

Reply:
This was my understanding of what @sergiusens intended - for us to be a pretty slim wrapper around the actual API. I think his intention was for us to pass the data structures as-is so that we don't need to update craft-store to be able to use additional properties that get added on the store side (either in the input or in the output).